### ISSUE TITLE
Boost Context doesn't compile when cross compiling

### ISSUE THREAD

**Issue by edtanous** [ruslo/hunter#553]
Minimum implementation that shows the issue.  

``` cmake
#CMakeLists.txt
cmake_minimum_required(VERSION 3.5)

include("HunterGate.cmake")

HunterGate(
    URL "https://github.com/ruslo/hunter/archive/v0.16.6.tar.gz"
    SHA1 "3058054dd0226692869b6e8c50fbb057562a54bd"
)

project (FOO C CXX)

hunter_add_package(Boost COMPONENTS context)
```

``` cmake
#arm.cmake
SET(CMAKE_SYSTEM_NAME Linux)
SET(CMAKE_SYSTEM_PROCESSOR "armv6")

SET(CMAKE_CROSSCOMPILING True)

SET(CMAKE_C_COMPILER arm-linux-gnueabi-gcc) 
SET(CMAKE_CXX_COMPILER arm-linux-gnueabi-g++)

set(triple arm-linux-gnueabi)
set(CMAKE_C_COMPILER_TARGET ${triple})
set(CMAKE_CXX_COMPILER_TARGET ${triple})
```

Execute with 

``` bash
cmake . -DCMAKE_TOOLCHAIN_FILE=arm.cmake
```

This configuration fails with an error that essentially boils down to the following in Boost.

```
gcc.compile.c++ bin.v2/libs/context/build/gcc-5.4.0/debug/link-static/threading-multi/unsupported.o
libs/context/src/unsupported.cpp:7:2: error: #error "platform not supported"
 #error "platform not supported"
```

Essentially, when cross compiling, boost cannot determine the ABI, architecture, and binary-format flags, so they need to be explicitly provided.  This is documented in boost here: http://www.boost.org/doc/libs/1_61_0/libs/context/doc/html/context/architectures/crosscompiling.html

Per the hunter documentation I tried to add these with the following code in a config.cmake

``` cmake
hunter_config(Boost VERSION 1.62.0
CMAKE_ARGS  
CONTEXT_abi=aapcs CONTEXT_binary-format=elf CONTEXT_architecture=arm
```

Unfortunately, the CONTEXT_abi argument results in the addition of "-s abi=aapcs" being added to the b2 arguments string.  This does does not have the desired effect, as the flag is one of the boost built-in flags and should not sit behind a "-s".  So far as I can tell, there is no way to add arguments that don't get prefixed with "-s".

Removing the -s from line 23 of hunter_boost_component_b2_args.cmake causes my above custom arguments to work as expected, and compilation of boost context to succeed, but this would cause a breaking change for others using the custom flags.

I'm happy to submit a patch, but I'm looking for some input on the desired resolution.  The two I see are:
1. Add a method of declaring boost built-in flags to the hunter_config.  Possibly prefixed by BOOST_.  This would mean I would add something like BOOST_architecture=arm to my CMAKE_ARGS to get the desired effect
2. If cross compiling, automatically add the architecture, abi, and binary-format flags to all b2 flags.  Getting these values from cmake isn't as straightforward as it seems, and could be problematic for non-cross compiled builds.

I'm also open to suggestion if there are other ideas.
##

**Comment by ruslo** [comment]
> Per the hunter documentation I tried to add these with the following code in a config.cmake

This information is available in toolchain already, we should not bother user with it.

> 2. If cross compiling, automatically add the architecture, abi, and binary-format flags to all b2 flags. Getting these values from cmake isn't as straightforward as it seems, and could be problematic for non-cross compiled builds.

That's the right approach. We should analyze the values of `CMAKE_SYSTEM_NAME`,  `CMAKE_SYSTEM_PROCESSOR`, `CMAKE_{C|CXX}_COMPILER_TARGET` and add corresponding flags to b2 configure step.

**Comment by edtanous** [comment]
The 3 pieces of info needed are: abi, binary-format, and architecture.
architecture can be determined from the CMAKE_SYSTEM_PROCESSOR argument.
So far as I can tell, cmake is not aware of the abi nor the binary-format in use.  To determine each of those, I'd likely have to generate a lookup table for all compilers defaults, as well as check for an explicitly declared -mabi flag.  I think that this large problem space is the same reason that boost went to forcing an explicit argument when cross compiling.
If you have suggestions, for how to determine abi and binary format, I'd love to hear them, but at this point, I'm thinking that option #1 above is my only option.

**Comment by ruslo** [comment]
> To determine each of those, I'd likely have to generate a lookup table for all compilers defaults, as well as check for an explicitly declared -mabi flag

What about [try_compile](https://cmake.org/cmake/help/latest/command/try_compile.html)? You can create small file with predefined macroses you're looking for and use the results to create b2 configure arguments, similar to toolchain-id calculation:
* https://github.com/ruslo/hunter/blob/master/scripts/ShowPredefined.cpp

**Comment by xsacha** [comment]
The Boost configuration is extremely messy. Especially because their jam scripts follow different cross-compiling paths depending on the toolset (compiler). I found this trying to use Android with clang.

It should be able to use the host clang to compile and that is no problem at all -- it does this. However, by default it tries to compile for host architecture as well. Manually setting my "  <architecture>" in the user jam didn't work. Boost detected this and ran the proper configs for it but never passed this on to clang. Also, the instruction-set does not let me pass on aarch64 (this is a 22-month issue that was never fixed).

Basically, there are many issues with the boost configuration which all need work upstream. I think the proper solution to this mess is fixing it at the source, which I am considering.

**Comment by ruslo** [comment]
Closing inactive issue with "Broken package" label, see details:
* https://github.com/ruslo/hunter/wiki/error.external.build.failed

**Comment by arcanis** [comment]
I just got hit with this issue - I'm trying to build a program using `coroutine2`, which depends on `context`.

Has a workaround been figured out yet?

**Comment by JasonDictos** [comment]
I've been battling this for some time, think its time to put on my big boy pants and submit a hunter pr once and for all ;p
